lightning/src/ln/channelmanager.rs

```diff
@@ -4284,6 +4286,61 @@ where
 			)
 	}
 
+	#[cfg(async_payments)]
+	fn initiate_async_payment(
+		&self, invoice: &StaticInvoice, payment_id: PaymentId
+	) -> Result<(), Bolt12PaymentError> {
+		let mut res = Ok(());
+		PersistenceNotifierGuard::optionally_notify(self, || {
+			let outbound_pmts_res = self.pending_outbound_payments.static_invoice_received(
+				invoice, payment_id, &*self.entropy_source, &self.pending_events
+			);
+			let payment_release_secret = match outbound_pmts_res {
+				Ok(secret) => secret,
+				Err(Bolt12PaymentError::UnexpectedInvoice) | Err(Bolt12PaymentError::DuplicateInvoice) => {
+					res = outbound_pmts_res.map(|_| ());
+					return NotifyOption::SkipPersistNoEvents
+				},
+				Err(e) => {
+					res = Err(e);
+					return NotifyOption::DoPersist
+				}
+			};
+
+			let reply_paths = match self.create_blinded_paths(
+				MessageContext::AsyncPayments(AsyncPaymentsContext::OutboundPayment { payment_id })
+			) {
+				Ok(paths) => paths,
+				Err(()) => {
+					self.abandon_payment_with_reason(payment_id, PaymentFailureReason::RouteNotFound);
+					res = Err(Bolt12PaymentError::SendingFailed(RetryableSendFailure::RouteNotFound));
+					return NotifyOption::DoPersist
+				}
+			};
+
+			let mut pending_async_payments_messages = self.pending_async_payments_messages.lock().unwrap();
+			const HTLC_AVAILABLE_LIMIT: usize = 10;
+			reply_paths
+				.iter()
+				.flat_map(|reply_path| invoice.message_paths().iter().map(move |invoice_path| (invoice_path, reply_path)))
+				.take(HTLC_AVAILABLE_LIMIT)
+				.for_each(|(invoice_path, reply_path)| {
+					let instructions = MessageSendInstructions::WithSpecifiedReplyPath {
+						destination: Destination::BlindedPath(invoice_path.clone()),
+						reply_path: reply_path.clone(),
+					};
+					let message = AsyncPaymentsMessage::HeldHtlcAvailable(
+						HeldHtlcAvailable { payment_release_secret }
+					);
+					pending_async_payments_messages.push((message, instructions));
```

Comment:
How should recipients handle messages with the same `payment_release_secret`?

Reply:
Recipients will just always reply with `ReleaseHeldHtlc` when they receive a `HeldHtlcAvailable` message, regardless of whether it's a duplicate, IIUC. They don't store any state relating to the payment until the keysend HTLC actually comes in.

Reply:
Ah, ok. So we'd essentially receive multiple invoices and de-dup as we normally do in `outbound_payments`. I guess that's the same as what we do for non-async payments.

Reply:
This is receive-side, so wouldn't be receiving multiple invoice in this case. But yeah we'd receive multiple identical `held_htlc_available` messages, reply to them all with `release_held_htlc` (since we don't have any state at this point so can't dedup), and then eventually receive a keysend payment.

Reply:
Ah, right, we already have the (static) invoice _here_, which is the send side *going to* the receive side with `held_htlc_available`.